pkg/codegen/pcl/binder_test.go

```diff
@@ -749,3 +749,84 @@ value = element(data, 0)
 	assert.Empty(t, diags, "There are no error or warning diagnostics")
 	assert.NotNil(t, program)
 }
+
+func TestBindingSelfReferencingComponentFailsWithCircularReference(t *testing.T) {
+	t.Parallel()
+	componentDir := filepath.Join(testdataPath, "self-referencing-components-pp")
+	files, err := os.ReadDir(componentDir)
+	if err != nil {
+		t.Fatalf("could not read test data: %v", err)
+	}
+	parser := syntax.NewParser()
+	for _, fileName := range files {
+		fileName := fileName.Name()
+		if filepath.Ext(fileName) != ".pp" {
+			continue
+		}
+
+		path := filepath.Join(componentDir, fileName)
+		contents, err := os.ReadFile(path)
+		require.NoErrorf(t, err, "could not read %v", path)
+
+		err = parser.ParseFile(bytes.NewReader(contents), fileName)
+		require.NoErrorf(t, err, "could not read %v", path)
+		require.False(t, parser.Diagnostics.HasErrors(), "failed to parse files")
+	}
+
+	var bindError error
+	var diags hcl.Diagnostics
+	absoluteProgramPath, err := filepath.Abs(componentDir)
+	if err != nil {
+		t.Fatalf("failed to bind program: unable to find the absolute path of %v", componentDir)
+	}
+
+	program, diags, bindError := pcl.BindProgram(parser.Files,
+		pcl.Loader(schema.NewPluginLoader(utils.NewHost(testdataPath))),
+		pcl.DirPath(absoluteProgramPath),
+		pcl.ComponentBinder(pcl.ComponentProgramBinderFromFileSystem()))
+
+	assert.Nil(t, program)
+	assert.NotNil(t, bindError)
+	assert.True(t, diags.HasErrors(), "There are error diagnostics")
+	assert.Contains(t, diags.Error(), "circular reference")
+}
+
+func TestBindingMutuallyDependantComponentsSucceeds(t *testing.T) {
```

Comment:
Do we have a test for the current error that circular deps between resources results in error? If not that needs adding.

Reply:
I'll check and add one if we don't

Reply:
Done ✅